contracts/ibc-reflect-send/src/ibc.rs

```diff
@@ -264,13 +274,15 @@ mod tests {
     // save the account (tested in detail in `proper_handshake_flow`)
     fn connect(mut deps: DepsMut, channel_id: &str) {
         // open packet has no counterparty version, connect does
-        let mut handshake_open = mock_ibc_channel(channel_id, IbcOrder::Ordered, IBC_VERSION);
-        handshake_open.counterparty_version = None;
+        let mut handshake_open =
+            IbcChannelOpenMsg::new(mock_ibc_channel(channel_id, IbcOrder::Ordered, IBC_VERSION));
+        handshake_open.channel.counterparty_version = None;
         // first we try to open with a valid handshake
         ibc_channel_open(deps.branch(), mock_env(), handshake_open).unwrap();
 
         // then we connect (with counter-party version set)
-        let handshake_connect = mock_ibc_channel(channel_id, IbcOrder::Ordered, IBC_VERSION);
+        let handshake_connect =
```

Comment:
This only works as long as `IbcChannelConnectMsg` contains no other information than the `IbcChannel`. This is the case right now but my understanding is other potentially required fields could be added. The whole point of this PR is to make clear `IbcChannelConnectMsg` can contain more than `IbcChannel`.

So in this particular case I have a tendency against implementing it.

Would it help to let the `mock_ibc_channel`/`mock_ibc_packet_recv`/`mock_ibc_packet_ack` testing helpers return the message instead of the plain ?

Reply:
You are very correct.

`mock_ibc_channel` / `mock_ibc_packet_recv` / `mock_ibc_packet_ack` helpers is a great idea though (and change function signature if another field is added)

Reply:
When it comes to the mock functions, what I'm struggling with is the `packet_ack` one.

```rust
#[cfg(feature = "stargate")]
pub fn mock_ibc_packet_ack<T: Serialize>(
    my_channel_id: &str,
    data: &T,
) -> StdResult<IbcPacketAckMsg> {
    let packet = IbcPacket {
        // packet fields
    };
    
    Ok(IbcPacketAckMsg::new(packet)) // this errors because the constructor needs IbcAcknowledgementWithPacket
```

I could change `IbcAcknowledgementWithPacket` to contain `IbcPacketAckMsg` rather than the other way around.

Or `mock_ibc_packet_ack` could accept an `IbcAcknowledgement` as one of the arguments.

I don't understand this conceptually well enough to make the call, I think.

Reply:
Looking at how `mock_ibc_packet_ack` is used, you see that the `IbcAcknowledgementWithPacket` is constructed in the calling code anyways. E.g.

```rust
            let packet = mock_ibc_packet_ack(CHANNEL_ID, br#"{}"#).unwrap();
            let ack = IbcAcknowledgementWithPacket {
                acknowledgement: IbcAcknowledgement::new(br#"{}"#),
                original_packet: packet,
            };
            call_ibc_packet_ack::<_, _, _, Empty>(&mut instance, &mock_env(), &ack)
                .unwrap()
                .unwrap();
```

So the helper could create the `packet`, the `IbcAcknowledgementWithPacket` and the `IbcPacketAckMsg`.

We then just need two different helpers for `ibc_packet_ack` and `ibc_packet_timeout`.

Reply:
Sure, but different calls create different acknowledgements. Does it matter?

See [this](https://github.com/CosmWasm/cosmwasm/blob/fbb9f30ae7b0cf4aaaea287bd59043e381b58fd0/contracts/ibc-reflect-send/src/ibc.rs#L383) and [this](https://github.com/CosmWasm/cosmwasm/blob/fbb9f30ae7b0cf4aaaea287bd59043e381b58fd0/contracts/ibc-reflect-send/src/ibc.rs#L292).

Reply:
The acknowledgement (one of the representations in `acknowledgement: IbcAcknowledgement::encode_json(&response).unwrap(),`) should become a new argument to the helper.

Reply:
This probably also warrants adding info about changed mocks in the changelog and migrating guide, doesn't it?

Reply:
At least in the CHANGELOG for the API breaking changes. Not so sure about MIGRATING as Ethan is probably the only one who wrote IBC contracts so for. We can always add those if needed.